ocaml/idl/datamodel_certificate.ml

```diff
@@ -44,6 +57,12 @@ let t =
       ; field ~qualifier:StaticRO ~lifecycle
           ~ty:(String) "fingerprint" ~default_value:(Some (VString ""))
           "The certificate's fingerprint / hash"
+      ; field ~qualifier:StaticRO ~lifecycle:[Published, rel_next, ""]
+          ~ty:(Ref _certificate) "signed_by" ~default_value:(Some (VRef null_ref))
```

Comment:
I think this is the part that you said that duplicates the information in the X509 chain.
Does signed_by also imply that the signature is a valid? What happens when the certificate expires, or is revoked by a CRL or OCSP responder? Perhaps we should treat it as signed_by=has a valid digital signature from the other certificate, and to leave expiration and revocation validation up to OpenSSL (which probably won't do revocation validation on its own though).

Reply:
Also note that certificates are not uniquely identified in a chain, it is sometimes possible to construct different trust paths starting from same end user certificate, see https://medium.com/@sleevi_/path-building-vs-path-verifying-the-chain-of-pain-9fbab861d7d6, we should also consider the situation where a single certificate has multiple possible issuers, and multiple paths (i.e. instead of treating it as a chain, treat it as a graph).
Or perhaps we put the burden on the user to choose one valid trust path, and then we store just that in the DB?

Reply:
> Perhaps we should treat it as signed_by=has a valid digital signature from the other certificate, and to leave expiration and revocation validation up to OpenSSL (which probably won't do revocation validation on its own though).

This is how it's intended to be used: there are already fields for validity dates and CRL are only used when users do an explicit API call, xapi does not use them on the background to achieve revocation. (instead certificates are remove from the trust root)

> we should also consider the situation where a single certificate has multiple possible issuers, and multiple paths

This is indeed an issue. Currently the pool level certificates are all trusted. Hosts may have intermediate certificates along the server certificate to have be able to validated using trusted certificates.

One of the ideas to model all the intermediate certificates was to alert users of about-to-expire certificates. I wanted to also model the difference between self-signed certificates and CA-issued ones, but this is already done by the field `_type`. I'm rethinking the need to model the paths within the database and how it relates to the needs for the feature.

Reply:
> we store a pool level certificate (which itself can chain up to a trusted root somewhere in our store), and then check that host level certificates are signed by that certificate, i.e. we only have 2 levels: the CA level and the issued host level, and no more levels?

This is the situation with the new proposal: `pool` certificates are trust anchors for all the hosts in the pool and can validate host certificates. `host` certificates are server certificate and identify a single host. They are validated by at least one pool certificate. Auto-generated certificates are `pool_and_host` because they are both trust anchors and identify a server.